Adopt Annual Release Cycle (DEP 20) (#109)
django/deps: draft/0020-annual-release-cycle.rst

================================
DEP 0020: Annual Release Cycle
================================

:DEP: 0020
:Author: Carlton Gibson
:Implementation Team: TBD
:Shepherd: TBD
:Status: Draft
:Type: Process
:Created: 2026-04-24

.. contents:: Table of Contents
   :depth: 3
   :local:

Abstract
========

This DEP proposes that Django adopt an annual release cycle for major
versions, replacing the eight-monthly cycle established in `DEP 4`__.

__ https://github.com/django/deps/blob/main/final/0004-release-schedule.rst

Under the proposed schedule, one major version of Django is released each
calendar year, in January, and is supported for three years: one year of
bugfix and security releases, followed by two years of security releases
only. At any given time, three major versions are in support. The existing
two-tier feature-release/LTS distinction is retired, with every release

> receiving what was previously the LTS-level commitment.

Reviewer: As someone who's spent years warning clients that LTS releases offer a false sense of safety, I welcome this change! There was never a clear upgrade path / unified changelog from one LTS to the next, so migrating between LTS releases was always A LOT more stressful than upgrading to every single release in between.

Having no distinction between the LST/non-LTS releases makes things much clearer and easier!

Django's version numbering moves to Calendar Versioning in the form
``YYYY.N``, so that the release year is encoded in the version itself.

The first release under this DEP is Django ``2028.1``, replacing what would
otherwise have been Django 7.0.

This DEP supersedes DEP 4 and subsumes the clarifications in `DEP 44`__.

__ https://github.com/django/deps/blob/main/final/0044-clarify-release-process.rst

Specification
=============

Release Cadence
---------------

Django will issue one major release per calendar year.

Final major releases will be made in approximately the second week of
January, carrying the new year's number. The exact release day remains at
the Releaser's discretion, following the factors noted in DEP 44.

Each major release is preceded by a pre-release phase:

* An alpha release in early October, following the corresponding CPython
  final release.

> * A beta release approximately four weeks before the release candidate.

Reviewer 2: Question: This puts the beta release just as people are likely ramping down work for Christmas (other holidays also available). Is there concern that the beta will receive less testing because of that?

Reviewer 3: We probably want a month between each:
- alpha: mid oct
- beta: mid nov
- rc: mid dec
- final: mid jan

So the beta will fall mid nov, not dec 1, and releasers can do the last RC before going "away" over holidays, and merge and backport any last minute bugs in early jan

Dropping the beta in mid nov seems fine, the flip side of work "ramping down" after november is that you have more "unexpected" time to play with the beta 😉

Reviewer 4: I also had a similar comment:
> If 2028.1 is released mid Jan, then 2028.rc1 (is that is what we will name it? or 2028.1.rc1) will be released 2 weeks before being roughly January 1st. I think that is logistically tough with Fellows vacation and likely vacation of translators. So perhaps 4 weeks before we do the release candidate, roughly mid December, 4 weeks before that the beta, roughly mid November, the alpha at the start of October

Reviewer 5: Is there any reason to wait for the new Python final release to publish Django's alpha?

Python's RC1 is early August and RC2 is early September so it's already very representative of what will ship.

How about this calendar?
- Alpha: mid September (so Django's CI can run against Python RC1 in August and against RC2 before publishing it)
- Beta: mid October (after Python's final release)
- RC: mid November
- Final: early December (if two weeks between RC and final)

That way, Django is even closer to Python's schedule and everything is wrapped by the time of the Christmas/New Year holidays.

(With such a calendar, the version released in December 2027 would be called 2028)

Reviewer 3: >  Is there any reason to wait for the new Python final release to publish Django's alpha?

I don't think there's a reason to wait for Python pre-releases, no. (We have a daily build against the in-development Python already.) Extrapolating from my experience last year, we'll get about a month's worth of "interesting" bugs relating to the new python into early November.

What I like about @rik's proposal:
- aligning even closer with the Python calendar. Under the "plus last yellow" plan, I think we still have Django support for an EOL python for about 3 months (from October to the end of the year). Your proposal shaves off a month there.

What I like less:
- dealing with a stream of release blockers over the holidays
- if we're going full "car versioning" instead of "cal versioning" by releasing "early", then let's just back up another month and release in November so that we can drop the last python as soon as it goes red at the end of October.

Reviewer 5: My example calendar was just illustrative. I'm happy if we find a sustainable way to ship closer to Python.

> * dealing with a stream of release blockers over the holidays

I'm not sure which holidays and which release you are referring to. @jacobtylerwalls Could you clarify?

Reviewer 3: I meant the regressions that come in during the first weeks after the final release. Given the relative lack of pre-release testing, this is the busiest period involving the release. Releasing in early December means this period will fall Dec 7-28 or so when we have less availability.

* A release candidate approximately two weeks before the final release.

Bugfix and security releases continue to be made at the beginning of each
month for each supported stable version, as per DEP 4 and DEP 44. In
exceptional circumstances, additional releases may be made outside the
schedule (for example, for a high-severity security issue or a packaging
error).

As per DEP 4, non-security releases may be held back if there are
unresolved release blockers. The Steering Council may vote to delay a
non-security release if it believes there is reason to do so.

Versioning
----------

> Django will use Calendar Versioning in the form ``YYYY.N``, where:

Reviewer 6: Maybe link to https://calver.org/?

```suggestion
Django will use `Calendar Versioning <https://calver.org/>`__ in the form ``YYYY.N``, where:
```

> * ``YYYY`` is the four-digit calendar year of the major release.

Reviewer 5: Starting a thread to gather the conversation on the major version format. I think of [@nessita's comment](https://github.com/django/deps/pull/109#pullrequestreview-4171628128) as "comment 0" of this thread.

I'm strongly in favour of using the year in the version number. I believe it greatly clarifies communication. It will probably help Django's users upgrade: it's easier to communicate to managers and other stakeholders that you need time to upgrade something called "Django 2023" than it is something called "Django 5".

I also believe it is great for Django's marketing. Many people not using Django believe it is old and does not receive meaningful updates. When those people read titles such as "Here's what's new in Django 2028", they get a more positive impression than reading "Here's what's new in Django 8".

> * ``N`` is a monotonically incrementing counter within that year, starting

Reviewer 7: I wonder if this could either start at `0` or be of the format `NNN`. The point here is not to create the assumption that this number represents the month in said year. Especially given that there is a monthly release. 

Unless of course we do want to closely match the month, but not quite if there is an urgent security release?

Author: And also later in the cycle (years 2 and 3) many months may pass between releases. Making the second number the month was rejected (via the Forum discussion IIRC) because it would leave a question as to "Did I miss one" in such a case. 

`NN` might work. 

My preference is for `.1` over `.0` — it's the number of the release in the cycle. 

---

General point: Discussing this over the last few months, it's become clear that the _exact_ numbering scheme is a "What shade?" question (vs falling into line with the Python release process, and removing the gap between LTSs, which are the substance) 

As such, I'm happy to go with whatever emerges as consensus on the numbering scheme. The text presented here captures what I **think** was the balance of opinion so far. (I use YY for the year in my own projects, and initially favoured that, for example.)

Reviewer 8: @carltongibson that would be different from Python, which starts patch releases with a `.0`. I know they don't use CalVer, but most CS engineers start counting at zero anyway.

“Skipping” the 0 might lead to confusion. I would certainly in my case.

Let's compare it to the biggest adopter of CalVar, that small 4 trillion dollar startup; they use iOS 26.0 (I hate the missing 20).

Lastly, 2028.1 would be the “first” patch for v2028. It feels easier to community care with a colleague in the office.

Author: Yeah, again, for 0-based or 1-based I’m happy to go with the consensus. As I had it “1” is the first release. “2” is the second release and so on. (1-based is quite normal for CalVer schemes, I think. I certainly use them that way elsewhere.) 

I think beyond the first viewing either way would be clear enough.

Reviewer 6: Another +1 for starting at 0.

[pip](https://pypi.org/project/pip/#history) and [packaging](https://pypi.org/project/packaging/#history) come to mind as starting at 0. [attrs](https://pypi.org/project/attrs/#history) starts at 1.

.0 better aligns with the "major release" wording used elsewhere; "``2028.1`` is the major release" is a bit confusing at first, this sounds like a patch release.

(As an aside, I try and use "feature release" for CPython 3.x over "major release" to avoid no-we're-not-semver confusion.)

> * ``N`` is a monotonically incrementing counter within that year, starting

Reviewer 6: Because there will be a 2028.13 in 2029 (or something like that), perhaps:

```suggestion
* ``N`` is a monotonically incrementing counter within that major release, starting
```

>   at ``1`` for the major release. Each subsequent bugfix release

Reviewer 4: Given the quantity of security reports, it is likely we will have a release like `2028.36`, I think that's fine but just wanted to highlight it

  increments ``N``.

For example:

* ``2028.1`` is the major release made in January 2028.

> * ``2028.2`` is the first bugfix release following ``2028.1``.

Reviewer 6: Perhaps include an example 2028.x that happens in 2029.

> * ``2029.1`` is the next major release, made in January 2029.

Reviewer 9: This is amazing, btw! 

The "within that year" wording is ambiguous and may cause confusion. (Unless I'm confused by your intent.) 

Maybe it's worth spelling out or showing an example that even if one of those patches in the series lands in 2029, it might still remain in 2028.N series. 

(updated to switch from `.x` to `.N` to match the proposed change)

Reviewer 10: Yes, this wasn't overly clear to me too. I presume a security fix released in 2029 would be applied to the 2028 release train as 2028.x?

Author: Thanks, I'll clarify. 

The idea is to the December of the third year. (I said this better elsewhere...)

Version numbers remain orderable under the standard comparison rules used
by Python packaging tools. No third number is used; there is no separate
"patch" component.

Support Window
--------------

Every major release is supported for three years from its date of release:

* **Year 1**: bugfix and security releases.
* **Years 2 and 3**: security releases only.

At any given time, three major versions are under support: the current
release and the two prior years' releases. On the release of
``YYYY.1``, bugfix support for ``(YYYY-1).1`` ends, and security support
for ``(YYYY-3).1`` ends.

Every release receives the same support commitment. The explicit "LTS"
label used in DEP 4 is retired; see Rationale.

Python Version Support
----------------------

> Each Django major release supports three Python versions: the two Python

Reviewer 2: Question: Do older Django versions receive support for newer Python versions? My reading is no, since supporting a newly-released Python version isn't a "bugfix [or] security release".

The downside of this being developers not only have to wait for 3 months before adopting newer Python versions, but they _also_ need to upgrade Django too.

Author: I need to flesh this out more... 

The idea is that the current version picks up the new Python whilst it's within its first year, but the older two versions would not. 

There's no **requirement** to update your Python immediately to the newest version released, as the Django you're on will support 4, and the new one will only drop the last (that your current version supports) — If that makes sense. 

I'll spell this out more clearly.

Reviewer 8: Hm… that's a lot of dinosaurs burned to test all versions. Especially since 3rd parties will have to support multiple Django versions. @tim-schilling has made a similar point previously. I'd like to hear his opinion too.

Personally, if a Python version doesn't receive bug fixes, any bugs like the garbage collection bug may need a backport in Django because an old Python version doesn't receive a patch.

People don't need to update neither Django nor Python for quite a while, wichout each Django version supporting a large Python version ragenes.

In short, I'd only support the 2 active Python versions at the time of release (without a hard pinned upper bound) so people can try a newer Python release.

Author: @codingjoe OK... so the alternative there would be a _Green Only_ policy. (Just the current two Pythons at release) which brings it down another one. ("Another" because the current policy has the old LTS supporting five full versions at EOL.) 

I'm not 100% anti that. I went _Plus last yellow_ because that seemed to be the community vibe. 

But this is a topic we can decide on, and probably settle in a giant Zoom call 😅

Reviewer 8: I can live with both, as I have been living with the current system. But in reality I believe there are two kinds of Django projects. The fossil that hasn't been updated since some agency wrote it and still runs on Django 1.8. And people who are on the prev. LTS but have an active ticket to upgrade.

Never have I experienced upgrading a Python version as a big consideration since Python 3. But mileage may vary.

versions with active upstream support at the start of the Django
pre-release phase ("green"), plus the most recent Python version to have

> moved to security-only or end-of-life status ("plus last yellow").

Reviewer 4: During the active support of bugfix and security releases for `2028.1` (actually should we just call this `2028` when referring to that release in general?) there will be a new Python released in the October of 2028 (Python 3.17), I assume we will backport support of this Python to `2028.1`?

respond to any deprecation before it is removed.

Transition
----------

Releases up to and including Django 6.x continue on the existing schedule
under DEP 4 and DEP 44. Django 5.2 LTS and Django 6.2 LTS receive their
full advertised support windows.

The first release under this DEP is Django ``2028.1``, replacing what
would otherwise have been Django 7.0 (scheduled under the existing cycle
for December 2027). ``2028.1`` is released in January 2028 and is
supported for three years on the terms above.

Motivation
==========

*Outline — WIP to be expanded*

* **Alignment with Python's annual release cycle.** Python has released
  annually since Python 3.9. Django's eight-month cycle is continually
  misaligned with Python's, complicating Python support decisions and
  increasing the matrix of supported interpreters.

* **Removing the gap between LTS cycles.** Under the current policy, users
  on an LTS release do not receive most bugfixes, which land only on the
  current feature release. There are reports of users ceasing to file
  issues with Django and instead maintaining private patches against
  their LTS version. Making every release an LTS closes this gap: every

>   supported user receives bugfixes for a full year.

Reviewer 3: Glad to see this here, I'm happy about the bugfix window widening from 8 to 12 months. (I'm always sad about the regressions that just barely miss the shot clock.)

  maintainers under pressure to support more versions at once, for
  longer.

* **Version and timing clarity (secondary).** The present ``X.Y``
  numbering mimics semantic versioning despite not being semver, and the
  numbers carry no information about when a release was made or how
  current it is. ``YYYY.N`` encodes both, and makes Django's actual
  stability and upgrade story easier to communicate. This is a genuine
  improvement but is not the primary driver for this DEP.

Rationale
=========

Annual cycle
------------

Python's move to annual releases has made the existing eight-month cycle
progressively harder to reason about. An annual cycle aligns Django's
pre-release phase with the CPython release in October, giving Django a
predictable window in which to validate against a newly released Python
before Django's own final release in January.

A January final release, carrying the new year's number, gives a
"New Year, not Last Year" feel to each release and avoids the December
holiday period for both the Releaser and users evaluating upgrades.

Annual releases are (of course) slower than eight monthly ones, and that is a

> cost or this proposal. At this stage in Django's lifecycle, and given its

Reviewer 11: Typo?
```suggestion
cost for this proposal. At this stage in Django's lifecycle, and given its
```

"Plus Last Yellow" is the less aggressive of the two. It gives users one
additional Python version per Django release, smoothing the Python
upgrade path and remaining more compatible with the support windows of
common Linux distributions, where a stable release typically tracks a
specific Python version for around three years.

Third-party package maintainers, particularly those of newer or
still-evolving packages, may reasonably adopt a "Green Only" Python
support policy for their own releases, in order to reduce their own
maintenance burden, without this being in conflict with Django's wider
policy.

Related work: automated upgrade tooling
---------------------------------------

Effective use of a yearly cadence benefits from good automated upgrade
tooling. The ``django-upgrade`` project provides fixers that rewrite user
code across Django versions. Formalising any role for such tooling is out
of scope for this DEP and is suitable material for a separate proposal.

Backwards Compatibility
=======================

API stability commitments under this DEP are preserved: deprecated
features continue to warn for at least two major releases before removal,

> and security support extends for three years from each release.

Reviewer 6: Another plus: not only are stability commitments preserved, but the deprecation period is longer now, giving people more time to react.

  and similar) continues to work, since ``YYYY.N`` remains a
  PEP 440-compliant, orderable version. Code that relies on ``X.Y``
  having a bounded major component, or that uses ad hoc string parsing,
  will need updating.

* **The "LTS" label.** External resources (hosting providers, tutorials,
  books, third-party packages) that refer to "the Django LTS" will need
  to be updated to reflect that every release is long-term supported.
  The Marketing Working Group and the documentation are expected to lead
  on this communication.

* **Release tooling and schedules.** Internal release tooling, the
  ``internals/release-process/`` documentation, and the release schedule
  published on djangoproject.com all need to be updated to reflect the
  new cadence and numbering.

* **Third-party package metadata.** ``Framework :: Django :: X.Y``
  trove classifiers, tox environments, and CI matrices in the wider
  ecosystem will need to be adjusted. Because the transition takes
  effect at a clearly signposted release (``2028.1``), maintainers have
  advance notice.

No change is proposed to the support commitments already made for Django

> 5.2 LTS or Django 6.2 LTS under the existing schedule.

Reviewer 3: I think 7.0 works. If I count right, making 6.2 the first annual release would shorten its feature development period from 8 to 5 months this time. Better to roll this out with a more supple set of features, no? ("Django took a whole year and did what exactly?")

References to, for example, Steering Council terms that are grounded in LTS

> cycles will need to be updated to simply state two-years, or as appropriate.

Reviewer 2: Suggestion: Given LTS release are 3 years, it might make sense to align to that? It means everything follows a 3-year cycle. Granted it does increase the term substantially which may not be ideal.

Author: My personal take is that two years on the SC is just about right. 

But 🤷

(Either way, out of scope for this DEP — the note here is only that other docs will need to adjust phrasing.)

schedule (for example, for a high-severity security issue or a packaging
error).

As per DEP 4, non-security releases may be held back if there are
unresolved release blockers. The Steering Council may vote to delay a
non-security release if it believes there is reason to do so.

Versioning
----------

Django will use `Calendar Versioning <https://calver.org/>`__ in the form
``YYYY.N``, where:

* ``YYYY`` is the four-digit calendar year of the feature release.
* ``N`` is a monotonically incrementing counter, starting at ``0`` for the
  feature release. Each subsequent patch release increments ``N``.

For example:

* ``2028.0`` is the feature release made in January 2028.
* ``2028.1`` is the first patch release following ``2028.0``.
* ``2028.24`` is the 24th patch release of the ``2028.x`` version, likely
  occurring late in the support window.
* ``2029.0`` is the next feature release, made in January 2029.

Version numbers remain orderable under the standard comparison rules used
by Python packaging tools. There is no third numeric component; patch
releases increment ``N``.

Support Window
--------------

Every feature release is supported for three years from its date of

> release:

Reviewer 12: This part of the proposal excites me the most as a Django user and contractor!

Most of my clients stick with LTS releases of Django because there is not always enough time/money/willpower to upgrade Django in time when a non-LTS release goes out of support. Hence the preference to stick with LTS releases for long term continuity and security support.

The consequence is a bigger upgrade jump, like from v4.2 to v5.2, which this proposal would also solve. We could just upgrade to the next release to gain another year of support. In practice, we would probably upgrade to the latest release soon after it comes out in January.

If I imagine for a moment Django would have had this policy, I would have been able to use features from non-LTS releases like `GeneratedField` and the `django.tasks` framework much sooner.

my 2 cents from the perspective of a Django user 👍

* **Year 1** is considered *mainstream support*, receiving fixes for security
  issues, data loss bugs, crashing bugs, major functionality bugs in
  newly introduced features, and regressions from older versions of Django.
* **Years 2 and 3** are considered *extended support*, receiving fixes for
  security issues and data loss bugs only.

The support level for a feature release changes in the December of each year,
at the point just after that month's releases are (or would be) made. At this
time, the current year's release enters extended support and the year-three
release becomes end-of-life (EOL). The new year's feature release then begins
mainstream support the following month.

This means that, bar the transition period over the new year, at any given time,
three feature releases are under support: the current release, under mainstream
support, and the two prior years' releases, under extended support.

Every release receives the same support commitment. The explicit "LTS"
label used in DEP 4 is retired; see Rationale.

Python Version Support
----------------------

> At it's release, each Django feature release will support three Python

Reviewer 6: ```suggestion
At its release, each Django feature release will support three Python
```

community as a whole.

At the same time, it's not realistic for us to offer more frequent releases,
whilst maintaining Django's support and stability guarantees. Nonetheless,
Django's main development branch is itself (remarkably) stable. Teams wanting
to be on the cutting edge have for many years deployed from the main branch.
Whilst out of scope for this DEP, it remains open for future work to improve
the messaging around testing and using Django's main development branch, and to
assess the feasibility of us providing incremental development releases as part
of the monthly release process.

Every release as LTS
--------------------

The current two-tier system has a well-documented failure mode: the
backport policy means that users on an LTS receive only a small subset of
fixes, while users who stay current receive bugfixes but face more
frequent upgrades. Collapsing this into a single tier — every release
supported for three years — removes the failure mode without reducing the
commitment offered to any user.

The explicit "LTS" label is retired in this DEP on the grounds that every
release now carries the commitment that label previously denoted. It

> remains the responsibility of the Marketing Working Group, the

Reviewer 13: The marketing working group isn't a formal WG and should probably be replaced by the SC and / or board.

There's another bit in the rationale that needs updating.

Author: f69fe337e5d0653013b3a9f43e9834836b76d42b
